### PR TITLE
Full review of farm - Issue 166

### PR DESCRIPTION
**Pull request by jdgarciauc3m**
Revised all documentation and code of farm.

Also added new example that capitalizes text.

### CODE REVIEW

File: doc/farm.md [hunk 1]
```diff
@@ -12,38 +14,59 @@ grppi::farm(exec, other_arguments...);
...
+given an input value `x` of type `T` makes valid the following:
+
+~~~{.cpp}
+U res{transformer(x)};
+~~~
+
+A *stand-alone farm* also has a **Generator**. The generator may be any C++
+callable entity that produces values in a way that allows to signal an
+indication of end-of stream. For this purpose, the farm requires that the
+generator produces values of any type that is convertible to `bool` and can
```
Comment (mdolz) [inline doc/farm.md:40]: Replace: `and can dereferenced` -> `which can be dereferenced`

File: doc/farm.md [hunk 2]
```diff
@@ -57,18 +80,13 @@ farm(exec,
...
-
-* A **Generator** of values.
-* An **Operation** performing a transformation on generated values.
-
-**Note:** In this case the unary operation results are discarded. Thus, the result type may be `void` or any other type.
+A *stand-alone identity farm* generates a stream of data items and then consumes
+them.
 
 ---
 **Example**
```
Comment (mdolz) [inline doc/farm.md]: Forgotten description for the example: `**Example:** Composed pipeline ...`

File: doc/farm.md [hunk 3]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
-
-* An **Operation** performing a transformation on generated values.
-
-The input values will be generated by the upper level pattern, which will be also repsonsible for consuming the output values.
+A *composable farm* applies a **Transformer** to each data item in a stream. The
+**farm** can be inserted into another upper level pattern (which will
+be responsible for generation and consumption)
 
 ---
 **Example**
```
Comment (mdolz) [inline doc/farm.md]: Forgotten description for the example: `**Example:** Composed pipeline ...`

File: doc/farm.md [hunk 4]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
-    },
+  farm(exec, [](auto x) {
+    return x.length();
+  }),
   stageC
-  );
+);
 ~~~
----
 **Note**: For brevity we do not show here the details of other stages.
```
Comment (mdolz) [inline doc/farm.md]: This note does not follow the same MD format as you used for the filter pattern.

File: doc/farm.md [hunk 5]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
 pipeline(exec,
   stageA,
   stageB,
   print_long_words,
   stageC
-  );
+);
 ~~~
----
 **Note**: For brevity we do not show here the details of other stages.
```
Comment (mdolz) [inline doc/farm.md]: This note does not follow the same MD format as you used for the filter pattern.

File: doc/farm.md [hunk 6]
```diff
@@ -12,38 +14,59 @@ grppi::farm(exec, other_arguments...);
...
-* A **Consumer** of generated values.
+A *stand-alone transforming farm* transforms each generated element in a stream
+and then consumes it.
 
 ---
-**Example**
+**Example**: Generate a sequence of integers and print the square of each value.
 ~~~{.cpp}
 int n = 10;
 farm(exec,
```
Comment (mdolz) [inline doc/farm.md]: Forgotten `grppi::` qualified name for the `farm` function.

File: doc/farm.md [hunk 7]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
-
-The input values will be generated by the upper level pattern, which will be also repsonsible for consuming the output values.
+A *composable farm* applies a **Transformer** to each data item in a stream. The
+**farm** can be inserted into another upper level pattern (which will
+be responsible for generation and consumption)
 
 ---
 **Example**
 ~~~{.cpp}
 pipeline(exec,
```
Comment (mdolz) [inline doc/farm.md]: Forgotten `grppi::` qualified name for the `pipeline` function.

File: doc/farm.md [hunk 8]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
   stageA,
   stageB,
-  farm(exec,
-    [](auto x) {
-      if (x.length() > 4) std::cout << x << std::endl;
-    },
+  farm(exec, [](auto x) {
```
Comment (mdolz) [inline doc/farm.md]: Forgotten `grppi::` qualified name for the `farm` function.

File: doc/farm.md [hunk 9]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
 For composing complex patterns, the `farm()` function may be used to create an object that may be used later in the composition.
 
 ---
 **Example**
 ~~~{.cpp}
-auto print_long_words = farm(exec,
-    [](auto x) {
-      if (x.length() > 4) std::cout << x << std::endl;
-    }
+auto print_long_words = farm(exec, [](auto x) {
```
Comment (mdolz) [inline doc/farm.md]: Forgotten `grppi::` qualified name for the `farm` function.

File: doc/farm.md [hunk 10]
```diff
@@ -81,55 +99,52 @@ farm(exec,
...
+  if (x.length() > 4) std::cout << x << std::endl;
+});
 
 pipeline(exec,
```
Comment (mdolz) [inline doc/farm.md]: Forgotten `grppi::` qualified name for the `pipeline` function.

**Review by mdolz — CHANGES_REQUESTED** [review]
Changes requested are inlined in the codes.

File: doc/farm.md [hunk 11]
```diff
@@ -81,55 +99,52 @@ farm(exec,
     if (file) return word;
```
Comment (javidoc00) [inline doc/farm.md:105]: In line 96 file variable should be passed by value or reference to the lambda function.

**Review by javidoc00 — CHANGES_REQUESTED** [review]
changes are inline

File: doc/farm.md (hunk 11 above)
Comment (jdgarciauc3m) [reply doc/farm.md:105]: It needs to be by reference.

File: doc/farm.md (hunk 1 above)
Comment (jdgarciauc3m) [reply doc/farm.md:40]: Solved by commit 223a37b6d630d856ce55831446c30e8e09403f75

File: doc/farm.md (hunk 11 above)
Comment (jdgarciauc3m) [reply doc/farm.md:105]: Solved by commit 223a37b6d630d856ce55831446c30e8e09403f75

**Review by mdolz — DISMISSED** [review]

**Review by Drioastorga — DISMISSED** [review]
Ok

**Review by javidoc00 — APPROVED** [review]

**Comment by jdgarciauc3m** [comment]
Closes issue #166

**Pull request merged**